### DISCUSSION TITLE
AudioOutput.h : 387 - Available Buffer 0 too small 64: requested: 128 -> increase the buffer size

### DISCUSSION THREAD

**Discussion by Knottis** [pschatzmann/arduino-audio-tools#1601] [Q&A] [answered]
### Now trying to run my little program that worked perfectly last summer, but after I ran the "Year's" update of the library, I'm put back to the start!.. 

In this little program, I use OutputMixer to send one signal out to the speaker (something which only works if there is a signal in and not otherwise!..). I have accepted that it is not possible to do this with this library. 
The program contains the **OutputMixer** and **AudioRealFFT** functions.

But after the update, I get the following error when you start streaming the guitar or spotify sound to the program!..

### ERROR:
[W] AudioOutput.h : 387 - Available Buffer 0 too small 64: requested: 128 -> increase the buffer size
[W] AudioOutput.h : 387 - Available Buffer 1 too small 64: requested: 128 -> increase the buffer size

**_PS: Everything worked perfectly before this year's update!_**

The main parts of the program selection:

```c

#include <AudioConfig.h>
#include <AudioTools.h>

#include <ESP_I2S.h>
#include <BluetoothA2DPSink.h>

#include <AudioLibs/AudioRealFFT.h> 

const uint8_t I2S_SCK = 32;    
const uint8_t I2S_WS = 33;     
const uint8_t I2S_SDOUT = 22; 
I2SClass OutI2S;

BluetoothA2DPSink a2dp_sink(OutI2S);  

// Apply Effects on an Input Stream
AudioInfo info(44100, 2, 16);                   
SineWaveGenerator<int16_t> sineWave(10000);     
GeneratedSoundStream<int16_t> Sound(sineWave); 

OutputMixer<int16_t> mixer(OutI2S, 2);  // Output mixer with 2 - outputs mixing to AudioKitStream

const int buffer_size = 128;            // <<<=====   Split up the output into small slices 2 * 128 = 256

uint8_t sound_buffer[buffer_size];


// A2DP Sink with output to FFT...
AudioRealFFT FFT_DATA;                 // or any other supported inplementation

#define NUM_LEDS 15                    // Number of (15)-LEDs used for the VU meter. 


auto FFT_Conf = FFT_DATA.defaultConfig();
FFT_Conf.length = 1024;
FFT_Conf.copyFrom(info);

a2dp_sink.set_raw_stream_reader(Write_Data_Stream);  // Register callback FFT
a2dp_sink.set_stream_reader(Read_Data_Stream, false);  


mixer.begin(buffer_size);              // setup Output mixer with min necessary memory

```

**Comment by pschatzmann** [comment]
Did you try to increase the buffer size ? 
You try to write 128 bytes into a buffer with a max size of 128 that already contains 64 bytes

**Comment by Knottis** [comment]
**No matter what you set, you get this error!..**

const int buffer_size = 4096;          

[W] AudioOutput.h : 387 - Available Buffer 1 too small 2048: requested: 4096 -> increase the buffer size

**Reply by pschatzmann** [reply]
I could reproduce your issue: A [correction has been committed](https://github.com/pschatzmann/arduino-audio-tools/commit/d47537cb300187d47991e9585ded180bcd0149b1).

**Comment by Knottis** [comment]
**Have now update the library!.**

The error message disappeared and I have set the value back to 128. **_But no sound gets through!.._**

**Comment by pschatzmann** [comment]
I tested with [this sketch](https://github.com/pschatzmann/arduino-audio-tools/blob/d47537cb300187d47991e9585ded180bcd0149b1/examples/examples-audiokit/streams-generator_outputmixer-audiokit/streams-generator_outputmixer-audiokit.ino#L2) and I also reduced the copy size to 128 and mixer buffer size with mixer.begin(128);  This was working w/o any issues.

If I remember right the issue was that mixer.begin(n) did allocate a buffer with n samples instead of n bytes . Did you try with mixer.begin(128 * 2) ?

**Reply by pschatzmann** [reply, marked as answer]
I also committed a correction that would allow an output if one channel is empty: In the test case one copy line commented out and the  mixer.flushMixer() uncommented

**Comment by Knottis** [comment]
### No Sound Out!

Have tried some values, but do not get sound through when using **OutputMixer!..**.
128; // 256:  //512;  //1024;  //2048; //4096;  

_const int buffer_size = **1024**;_   

-------------
Or is the _OutputMixer_ functionality deleted from the library and this is _no longer supported_ in the new library version?.

**Comment by pschatzmann** [comment]
Can you provide a simple example that demonstrates the issue ?
All my test cases are still working...

**Comment by Knottis** [comment]
**You can only try your own code!..** 
Insert that you will use I2S output. Then you see that the Arduino goes into panic mode!..

_You cannot use your MacBook development machine for this. Install VritualBox with a clean machine with only Arduino installed and your library. Then you will quickly see what errors we users/amateurs experience from errors!_

_You really need to perform a quality check and cleanup of your code!._ 

https://github.com/pschatzmann/arduino-audio-tools/blob/185272cf19eed05e4e68c6bd32481883ddcbce31/examples/examples-communication/a2dp/basic-a2dp-mixer-i2s/basic-a2dp-mixer-i2s.ino


const uint8_t I2S_SCK = 32;    
const uint8_t I2S_WS = 33;     
const uint8_t I2S_SDOUT = 22;  
I2SClass OutI2S;

BluetoothA2DPSink a2dp_sink(OutI2S);

**Reply by pschatzmann** [reply]
As I was writing all my test cases including this one are passing when I tested with the Arduino ESP 3.0.1.  

The line BluetoothA2DPSink a2dp_sink; in the example is ok because the output is not used and is going via the callback.
The only negative thing I noticed was the when you hit pause in Bluetooth and the output is stopped the new i2s driver does not stop the output. But is a feature of the new IDF functionality.

What Arduino ESP32 core version are you using ?

Don't replace I2SStream with I2SClass!
 If you change the example, I expect that you should know what you are doing: Your changes should also give some compile errors!
